PEP-0496: Environment Markers

jamezpolley opened the pull request:
This started as the existing verbiage from PEP-0426 molded into a PEP-12 shape.

I've updated some of the wording around inequality comparison operators (the operators were listed but the text still claimed they couldn't be used), as well as splitting the markers into two types - version numbers and strings. I've added some description around how version numbers will be handled (in short, parsed as per PEP-0440)

The BNF needs to be updated. I was doing the updates on a beach over the weekend and decided to leave that bit of revision for a weekday.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/pypa/interoperability-peps/29)
<!-- Reviewable:end -->

ncoghlan commented:
Thanks for kicking this off - once you get those initial updates you mention done, I can assign this a number by putting a copy in the main PEPs repo.

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+And here's an example of some conditional metadata for a distribution that
+requires PyWin32 both at runtime and buildtime when using Windows::
+
+    "name": "ComfyChair",
```
So this metadata is in PEP-426 format which is still 95% hypothetical - there's no implementation of run_requires, for instance. I'd rather we used the actual formats that are around and work - requirements.txt format or setup_tools kwargs format. When I first read this I had to read it maybe three times to be sure I understood, email the list to check, and find out I didn't.

vsajip replied:
Actually, there *is* an implementation of `run_requires` in [`distlib`](https://distlib.readthedocs.org/en/latest/tutorial.html). Surely the answer is for `setuptools` to grow support for the new formats, rather than keeping the syntax as it is just because `setuptools` doesn't support it *now*. After all, it's not as if supporting the new syntax is especially hard.

rbtcollins replied:
Its not in use anywhere. And that syntax isn't end user syntax, its a proposed metadata format nothing uses - in practice. Aiming for comprehension by using existing things is a valuable approach here IMNSHO.

ncoghlan replied:
I concur with Robert here - as with PEP 440, the goal of pulling this out to a separate PEP is to make it independent of the other changes in PEP 426. That way, setuptools and pip can adopt it as a standardisation and slight modification of their *existing* capabilities, before we start reaching for the new capabilities potentially on offer through metadata 2.0.

ncoghlan replied:
I'd suggest using the current pip requirements file format as the source of examples for this PEP: https://pip.pypa.io/en/latest/reference/pip_install.html?highlight=markers#requirement-specifiers

jamezpolley replied:
I'm not entirely comfortable with solely using the requirements file format for this PEP. We already have two^Wthree implementations with slightly different syntaxes - the format pip reads in requirements.txt, the format pbr reads in setup.py, and the format setuptools accepts as kwargs.

Even assuming we ignore pbr as being its own extra thing unrelated to the PEP, I'd still prefer to either explicitly show both formats; or that we simply focus on the part that doesn't vary (ie, "sys_platform == 'win32'") and document elsewhere how to embed an environment marker into the various places they can be used.

I think I prefer the former approach - or at least, to focus the document on the invariant part and then maybe have a section at the bottom showing markers embedded in requirements.txt and setup.py (and pbr's docs can document how to put a marker into setup.cfg)

I note that https://pip.pypa.io/en/latest/reference/pip_install.html?highlight=markers#requirement-specifiers says "Don't use quotes in requirement files."  I've updated the requirements.txt examples here to not use quotes.

Given that setuptools doesn't actually support the syntax I'm documenting here, it sounds like it would be worth starting on an implementation of that as part of the implementation of this PEP? Once we have a working markerlib (or replacement) I imagine it won't be too difficult for setuptools to notice a marker in the requirements and pass it off to markerlib for evaluation (he says, with an abundance of optimism probably born of not having looked too closely at the code yet)

ncoghlan replied:
That's not what the note in the pip docs is referring to - it's referring to quoting entire version *specifiers*, not the individual values in environment markers.

You can see this in the example in the pip docs, where the Python version being compared against is quoted: `SomeProject ==5.4 ; python_version < '2.7'`

setuptools and pip both use the PEP 345 environment marker syntax as written, and this PEP will still *only* specify the environment marker syntax. However, the additional suggestion here is to switch to using the requirements.txt format in the *examples*, to help it make a bit clearer why environment markers can be useful.

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+Micro-language
+==============
+
+The micro-language behind this is a simple subset of Python. It compares::
```
I think we should remove 'simple subset'. Its not if we bring in ~=, and it being a subset of Python while a neat hack is not actually terribly meaningful or useful: an interpreter for this is a few dozen LOC with e.g. Parsely.

ncoghlan replied:
Right, that wording is just a holdover from the original PEP 346 design, and something we've now decided we want to change to align with PEP 440 for version numbering.

jamezpolley replied:
I've gone with "as follows"

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+* version numbers with the ``<``, ``<=``, ``>=``, and ``<`` operators
+  in addition to those supported for strings
+
+The usual boolen operators ``and`` and ``or`` can be used to combine
```
typo: boolean

jamezpolley replied:
done

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+The usual boolen operators ``and`` and ``or`` can be used to combine
+expressions, and parentheses are supported for grouping.
+
+The pseudo-grammar is not ::
```
not?

jamezpolley replied:
Is now!

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+    EXPR: ("(" MARKER ")") | (SUBEXPR [CMPOP SUBEXPR])
+    CMPOP: (==|!=|<|>|<=|>=|in|not in)
+
+but my memory of BNF isn't quite up to fixing it just now.
```
I suggest adding XXX or FIXME or something greppable where you expect to make edits.

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+
+but my memory of BNF isn't quite up to fixing it just now.
+
+``SUBEXPR`` is either a Python string (such as ``'2.4'``, or
```
Here this probably wants to say Python string, or PEP-440 version number, or marker variable.

Since we're revising this I have to question the utility of the quotes. They are useful if the interpreter is 'python itself', but if we write a small interpreter ourselves there's no ambiguity, and little value from the quotes.

ncoghlan replied:
In this case, the main argument for the quotes is "that's what PEP 345 used, and we're aiming to standardise and clarify an existing practice, rather than change it".

The history independent rationalisation is that it helps create a greater visual distinction from the preceding version specifier in the "specifier; marker" syntax that both pip and setuptools use.

jamezpolley replied:
Gah. But here the quotes are specified? but this should be ignored in requirements.txt?

I vote for taking them out; PEP 345 might have used them, but pip currently tells people not to use them. I'm going to vote for sticking with real-world practice rather than theoretical documentation.

ncoghlan replied:
As noted above, the comment about quotes in the pip docs is unfortunately placed, as it has nothing to do with the quote characters in environment markers, only with quoting entire version specifiers.

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+
+If a particular value is not available (such as the ``sys.implementation``
+subattributes in versions of Python prior to 3.3), the corresponding marker
+variable MUST be considered equivalent to the empty string.
```
This needs rephrasing I think, because '' is not terribly well defined in pep-440 (in fact, its illegal)

ncoghlan replied:
Yes, the PEP 440 equivalent is `'0'`

jamezpolley replied:
In fact I think it's bit more complex than that, since we now handle two kinds of variables. I've tried to clarify this for both strings and versions

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+lists on the right side of the ``in`` and ``not in`` operators,
+regardless of whether the comparison is with a version number, string,
+or one of the marker variable names (which either refers to a string
+or a variable number)
```
I think this can go if we remove the 'subset of Python' framing higher up. It becomes clear, like 'argparse>=2.3,<3' is not valid Python, that neither is this.

ncoghlan replied:
+1

jamezpolley replied:
Agreed and done

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,184 @@
+    implementation_version = format_full_version(sys.implementation.version)
+
+``python_full_version`` will typically correspond to the leading segment
+of ``sys.version()``.
```
This seems incompletely specified.

ncoghlan replied:
In what respect? The fact we only pay attention to a subset of PEP 440's scope for Python and implementation version numbers?

We should probably cross-reference https://docs.python.org/3/library/sys.html#sys.version_info and https://docs.python.org/3/library/sys.html#sys.implementation in relation to those additional constraints.

rbtcollins replied:
>>> import sys
>>> sys.version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'str' object is not callable


The specification is illegal. At minimum it means noone tested it.

>>> sys.version
'2.7.8 (default, Jun 18 2015, 18:54:19) \n[GCC 4.9.1]'

Has nothing defined as a 'segment' . Is it the first line? The version number leading up to the first space? ... ?

jamezpolley replied:
I've tweaked this to use sys.version and sys.version_info, and to give an actual working python definition of python_full_version.

I'm not sure that these changes have actually clarified things.

ncoghlan replied:
I don't think that paragraph is needed at all, as it's redundant text that apparently confuses rather than enlightens.

jamezpolley commented:
So I've done a few things to be more clear about the fact that we now support two types of variables - strings *and* version numbers, rather than just strings. In the main I think this has muddied the document as so many parts of it now say roughly a "Do this for strings, do this other thing for versions"

I think I'd like to make another pass to see if it becomes clearer if we put all the string information together, and all the version information together.

ncoghlan commented:
For merging, I'll likely dump the file from the PR, add it to the python.org PEPs repo to allocate it a number, and then commit it back here directly.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+Title: Environment Markers
+Version: $Revision$
+Last-Modified: $Date$
+Author: James Polley <jp@jamezpolley.com>
```
Header to add:

    BDFL-Delegate: Nick Coghlan <ncoghlan@gmail.com>

jamezpolley replied:
Done.

A few lines further down on the Replaces: line, I wasn't sure how to say that this is only a partial replacement for PEP-0345; is there a better way to do that?

ncoghlan replied:
I wouldn't worry about the Replaces line in this case - PEP 426 will eventually replace 345 entirely.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+
+For many packages this means they aim to support a wide range of
+Python releases. If they depend on libraries such as argparse or
+simplejson - which started as external libraries, but later got
```
simplejson is a bad example - it kept going as an independent project, and json is in the standard library for all the versions folks still care about. argparse is similarly available in all the versions folks care about.

Better to use the backport libraries like unittest2 or contextlib2 as examples, where folks want to use the backported 3.x version on earlier releases, but the standard library version on 3.x+

rbtcollins replied:
unittest2 should be used on all versions of Python - it has improvements from 3.6 backported to everything-else. Using the standard library version on 3.x is a poor idea and just makes using features harder. unittest2 <- the 2 is as in 'greater than 1', not 'for python 2.x'.

jamezpolley replied:
So it sounds like argparse is uncontroversial, so I'll trim simplejson.

I'll tweak the cross-platform paragraph as well to list pywin32 as an example of a platform-specific requirement, to emphasise that this isn't just about versions.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+difficult, as the set of required packages differs depending on the
+version of Python in use.
+
+For other packages, this means broad cross-platform support. However,
```
Suggesting rephrasing for clarity: "For other packages, designing for portability means supporting multiple operating systems. However, the significant differences between them may mean that particular dependencies are only needed on particular platforms (relying on `pywin32` only on Windows, for example)".

jamezpolley replied:
Oh hey, that's practically the same as what I wrote, but slightly cleaner. I'm going to steal it, thanks :)

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+
+Here are some examples of such markers inside a requirements.txt::
+
+   "sys_platform == 'win32'"
```
pywin32 >= 1.0; sys_platform == 'win32'

jamezpolley replied:
done

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+Here are some examples of such markers inside a requirements.txt::
+
+   "sys_platform == 'win32'"
+   "platform_machine == 'i386'"
```
Let's drop this example - depending on architecture is strange for Python apps.

rbtcollins replied:
It makes sense in the context of some C accelerated things, but I agree its not very relevant.

jamezpolley replied:
Dropped

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+
+   "sys_platform == 'win32'"
+   "platform_machine == 'i386'"
+   "python_version == '2.4' or python_version == '2.5'"
```
unittest2 >= 1.0; python_version < '3.5'

rbtcollins replied:
please no. Thats against upstream's advice on usage.

jamezpolley replied:
I'm not adding this in yet, but @rbtcollins do you have any docs that clarify this? I wasn't able to find anything in the docs in the unittest2 repo

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+   "sys_platform == 'win32'"
+   "platform_machine == 'i386'"
+   "python_version == '2.4' or python_version == '2.5'"
+   "python_version>='2.7' and python_version<'3.0'"
```
This example is aimed at demonstrating environment marker boolean logic, so a better example might be:

    backports.ssl_match_hostname >= 3.4; python_version < '2.7.9' or (python_version >= '3.0' and python_version < '3.4')

jamezpolley replied:
Done

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+   "platform_machine == 'i386'"
+   "python_version == '2.4' or python_version == '2.5'"
+   "python_version>='2.7' and python_version<'3.0'"
+   "'linux' in sys_platform"
```
More meaningful example:

    hawkey; 'linux' in sys_platform

jamezpolley replied:
would that be https://github.com/rpm-software-management/hawkey or https://pypi.python.org/pypi/hawkey ?

ncoghlan replied:
I'd assumed they were the same thing, but I just realised I'm not actually sure about that. I filed an RFE with the hawkey folks to make the status more explicit: https://bugzilla.redhat.com/show_bug.cgi?id=1244628

ncoghlan replied:
procfs would be a good example of an upstream Linux specific Python library: https://pypi.python.org/pypi/procfs/

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+
+* strings with the ``==`` and ``in`` operators (and their opposites)
+* version numbers with the ``<``, ``<=``, ``>=``, and ``<`` operators
+  in addition to those supported for strings
```
+1 I like this split.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+``Version number`` marker variables listed below. Comparisons between
+version numbers are done using PEP-0440 semantics.
+
+Version numbers
```
I suggest moving this header and list of variable names from here to...

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+in versions of Python prior to 3.3), the corresponding marker
+variable MUST be considered equivalent to the empty string.
+
+If a particular version number value is not available (such as
```
... here, such that the description of version variables comes entirely after the description of string variables.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+corresponding marker variable MUST be considered equivalent to ``0``
+
+
+Version information variables
```
With the move suggested above, this header can go away and the contents merged into the previous subheading.

jamezpolley replied:
Nice.

ncoghlan commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,174 @@
+    python_full_version = format_full_version(sys.version_info)
+    implementation_version = format_full_version(sys.implementation.version)
+
+``python_full_version`` will typically correspond to ``sys.version.split()[0]``.
```
Ah, nice :)

ncoghlan commented:
The general concept of the latest draft looks good to me, but I had several suggestions for the heading structure and the specific choices of examples.

jamezpolley commented:
I've put up some draft code over at https://bitbucket.org/pypa/setuptools/pull-requests/141/first-draft-of-an-implementation-of-pep/diff

jamezpolley commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+Examples
+========
+
+Here are some examples of such markers inside a requirements.txt::
```
Suggestion: adding some more explanatory notes to these examples ("Here is a requirements file that installs pywin32 on windows ...."

jamezpolley commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+
+   pywin32 >=1.0 ; sys_platform == 'win32'
+   unittest2 >=2.0,<3.0 ; python_version == '2.4' or python_version == '2.5'
+   backports.ssl_match_hostname >= 3.4 ; python_version < '2.7.9' or (python_version >= '3.0' and python_version < '3.4')
```
This could be written more cleanly as:

backports.ssl_match_hostname >= 3.4 ; python_version < '3.4' and python_version != 2.7.10

however that means we don't have a demonstrating of using parens. I think it's probably more important to give a readable example though.

jamezpolley commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+    implementation_version = format_full_version(sys.implementation.version)
+
+``python_full_version`` will typically correspond to ``sys.version.split()[0]``.
+
```
I'd like to add an "Alternatives Considered" section here that talks about some alternative approaches - such as having separate requirements files for each version of python, as OpenStack has tried - and why they have been rejected.

rbtcollins replied:
That would be fine, but the fundamental thing is covered by my proposed abstract lead-in.

jamezpolley commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+
+* ``python_version``: ``platform.python_version()[:3]``
+* ``python_full_version``: see definition below
+* ``platform_version``: ``platform.version()``
```
This actually returns a string:

>>> print platform.version()
Darwin Kernel Version 14.5.0: Wed Jul 29 02:18:53 PDT 2015; root:xnu-2782.40.9~2/RELEASE_X86_64

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+
+Abstract
+========
+
```
I think this starts too deep. If it started with something like the below it might be nice.

Packages need to have dependencies that are conditional on the environment the package is being installed into. Environment markers provide a small language for describing whether or not a particular dependency will apply to a given environment.

rbtcollins commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+environment. They are used to indicate when certain dependencies are only
+required in particular environments, and to indicate supported platforms
+for distributions with additional constraints beyond the availability of a
+Python runtime.
```
The last clause doesn't make sense to me. Where is that done?

jamezpolley replied:
I've reworded this, so I think it should be clearer now.

I think the "additional constraints" are things like sys.platform, which indicates windows/darwin/linux

jamezpolley commented on pep-0XXX-environment-markers.rst:
```diff
@@ -0,0 +1,171 @@
+* ``platform_release``: ``platform.release()``
+* ``implementation_name``: ``sys.implementation.name``
+* ``platform_machine``: ``platform.machine()``
+* ``platform_python_implementation``: ``platform.python_implementation()``
```
There seems to be some overlap here. For instance, platform_python_implementation is 'CPython' while 'implementation_name' is 'cpython'

I'm suggest that we list implementation_name and platform_version as being deprecated; or that we set implementation_name to lower(platform_python_implementation) if sys.implementation.name doesn't exist; or perhaps both at once.

platform_version needs to be listed as a string - it is not a version number.

    In [8]: platform.version()
    Out[8]: 'Darwin Kernel Version 14.5.0: Wed Jul 29 02:18:53 PDT 2015; root:xnu-2782.40.9~2/RELEASE_X86_64'

jamezpolley replied:
It is probably useful to link to the docs for the platform module to explain what these values return, or maybe to give a sample value for each, or perhaps just give a simple description. Otherwise it's not clear what we might expect in  "platform_machine" vs  "platform_release". It's trivial to find out but it'd be nice if people could tell just by reading the PEP without needing to go and run bits of code

jamezpolley replied:
platform has a dist() which returns a tuple containing the distribution name, version, and id on linux - empty strings on other platforms. This would allow for adding dependencies based on the version of ubuntu, or ubuntu vs redhat

ncoghlan commented:
The current draft of this is now included in the main PEPs repo as PEP 496, and in this repo as pep-0496-environment-markers.rst.

jamezpolley commented:
I've updated this with feedback from people at PyCon-AU

Notable changes:

* Abstract has been expanded
* Alternatives Considered section added
* backports.ssl_match_hostname example has been simplified - so there's no longer an example that uses parentheses, but that example is much more readable
* Each marker variable now has some sample values, so that a reader can get a clearer sense of what each does just from reading the PEP. To accomodate this I've changed that section from a list into a table. I've got examples from python3.4 on ubuntu, jython2.7, and python2.7 on OS X. Where a variable returns an empty string on a platform I've left it out.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -48,14 +50,32 @@ Environment Markers attempt to provide more flexibility in a list of
+Alternatives considered
+=======================
+
+For a while, OpenStack used seperate requirements files for different
```
I think the reference to OpenStack here needs explaining or removing.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -48,14 +50,32 @@ Environment Markers attempt to provide more flexibility in a list of
+``requirements-py3.txt``. However, in most cases the two files were
+near duplicates, with only one or two lines different; but maintaining
+two seperate files was more than twice as much work, as it required
+manual checking to keep the files in sync.
```
That wasn't the problem OpenStack had. The problem was building universal wheels was impossible. Also, markers predate OpenStack's efforts in this space, this PEP is updating the definition to be more useful based on lessons learnt - e.g. dealing with 2.7.10.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -48,14 +50,32 @@ Environment Markers attempt to provide more flexibility in a list of
+manual checking to keep the files in sync.
+
+Keeping all the requirements together in one place, with markers
+indicating which to install, has been easier to maintain.
```
This is a useful thing we've learnt, but I don't really see it as an alternative.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -48,14 +50,32 @@ Environment Markers attempt to provide more flexibility in a list of
 
-Here are some examples of such markers inside a requirements.txt::
+Here is an example of such markers inside a requirements.txt
+file. This will install ``pywin32`` on Windows; ``unittest2`` on
```
I still very very very very very much object to unittest2 being used conditionally. Its inappropriate. Please pick something where a conditional dependency is relevant - e.g. argparse. Also I suggest not using versions of Python < 2.6 in your examples.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -48,14 +50,32 @@ Environment Markers attempt to provide more flexibility in a list of
    pywin32 >=1.0 ; sys_platform == 'win32'
    unittest2 >=2.0,<3.0 ; python_version == '2.4' or python_version == '2.5'
-   backports.ssl_match_hostname >= 3.4 ; python_version < '2.7.9' or (python_version >= '3.0' and python_version < '3.4')
+   backports.ssl_match_hostname >= 3.4 ; python_version < '3.4' and python_version != '2.7.10'
```
So this example will be broken when 2.7.11 comes out. I think thats undesirable.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -73,8 +93,8 @@ Micro-language
 The micro-language behind this is as follows. It compares:
 
 * strings with the ``==`` and ``in`` operators (and their opposites)
```
Rather than saying opposites, lets spell them out. != and 'not in' I presume.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -73,8 +93,8 @@ Micro-language
 * strings with the ``==`` and ``in`` operators (and their opposites)
-* version numbers with the ``<``, ``<=``, ``>=``, and ``<`` operators
-  in addition to those supported for strings
+* version numbers with the ``==``, ``!=``, ``<``, ``<=``, ``>=``, and ``<`` operators.
```
what about ~= and === ?

jamezpolley replied:
Ho hum.

Haven't tested. Depends how those things are implemented. I suspect - since they're not standard python operators, but must (i assume) be specially handled in the code that parses dependencies - that they won't work. I'm not sure whether it's worth explicitly stating this: as far as I can recall they only have meaning when referring to python names, which we aren't doing here.

rbtcollins replied:
They are specifically for versions, which is the data type here. See PEP-440.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -100,30 +120,70 @@ version numbers are done using PEP-0440 semantics.
 Strings
 -------
 
-* ``os_name``: ``os.name``
```
btw the grammar docs reference SUBEXPR but its not in the grammar

jamezpolley replied:
SUBEXPR (line 112) should read STREXPR

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -100,30 +120,70 @@ version numbers are done using PEP-0440 semantics.
+   * - ``platform_dist_version``
+     - ``platform.dist()[1]``
+     - ``14.04``
+   * - ``platform_dist_id``
```
is dist_id made up? LSB releases' Distributor ID is e.g.'Ubuntu'. The value in this field is the LSB codename.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -100,30 +120,70 @@ version numbers are done using PEP-0440 semantics.
+If a particular string value is not available (such as
+``sys.implementation.name`` in versions of Python prior to 3.3, or
+``platform.dist()`` on non-Linux systems), the corresponding marker
+variable returned by setuptools will be set to the empty string.
```
by setuptools? That presumes implementation. Just say the variable will evaluate to the empty string.

rbtcollins commented on pep-0496-environment-markers.rst:
```diff
@@ -141,6 +201,10 @@ respectively, in accordance with the following algorithm::
 
+If a particular version number value is not available (such as
+``sys.implementation.version`` in versions of Python prior to 3.3) the
+corresponding marker variable returned by setuptools will be set to ``0``
```
likewise here.

ncoghlan commented:
Closing this PR, as environment markers ended up being incorporated into PEP 508: https://www.python.org/dev/peps/pep-0508/#environment-markers
